internal/component/database_observability/mysql/collector/schema_table.go

```diff
@@ -0,0 +1,227 @@
+package collector
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/grafana/loki/v3/pkg/logproto"
+	"github.com/hashicorp/golang-lru/v2/expirable"
+	"github.com/prometheus/common/model"
+
+	"github.com/grafana/alloy/internal/component/common/loki"
+	"github.com/grafana/alloy/internal/runtime/logging/level"
+)
+
+const (
+	OP_SCHEMA_DETECTION = "schema_detection"
+	OP_TABLE_DETECTION  = "table_detection"
+	OP_CREATE_STATEMENT = "create_statement"
+)
+
+const (
+	selectSchemaName = `
+	SELECT
+		SCHEMA_NAME
+	FROM
+		information_schema.schemata
+	WHERE
+		SCHEMA_NAME NOT IN ('mysql', 'information_schema', 'performance_schema', 'sys')`
+
+	selectTableName = `
+	SELECT
+		TABLE_NAME,
+		CREATE_TIME,
+		ifnull(UPDATE_TIME, CREATE_TIME) AS UPDATE_TIME
+	FROM
+		information_schema.tables
+	WHERE
+		TABLE_SCHEMA = ?`
+
+	// Note that the fully qualified table name is appendend to the query,
+	// for some reason it doesn't work with placeholders.
+	showCreateTable = `SHOW CREATE TABLE`
+)
+
+type SchemaTableArguments struct {
+	DB              *sql.DB
+	CollectInterval time.Duration
+	EntryHandler    loki.EntryHandler
+	CacheTTL        time.Duration
+
+	Logger log.Logger
+}
+
+type SchemaTable struct {
+	dbConnection    *sql.DB
+	collectInterval time.Duration
+	entryHandler    loki.EntryHandler
+	// Cache of table definitions. Entries are removed after a configurable TTL.
+	// Key is a string of the form "schema.table@timestamp", where timestamp is
+	// the last update time of the table (this allows capturing schema changes
+	// at each scan, regardless of caching).
+	// TODO(cristian): allow configuring cache size (currently unlimited).
+	cache *expirable.LRU[string, tableInfo]
+
+	logger log.Logger
+
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+type tableInfo struct {
+	schema     string
+	tableName  string
+	createTime time.Time
+	updateTime time.Time
+	createStmt string
+}
+
+func NewSchemaTable(args SchemaTableArguments) (*SchemaTable, error) {
+	return &SchemaTable{
+		dbConnection:    args.DB,
+		collectInterval: args.CollectInterval,
+		entryHandler:    args.EntryHandler,
+		cache:           expirable.NewLRU[string, tableInfo](0, nil, args.CacheTTL),
+		logger:          args.Logger,
+	}, nil
+}
+
+func (c *SchemaTable) Start(ctx context.Context) error {
+	level.Debug(c.logger).Log("msg", "SchemaTable collector started")
+
+	ctx, cancel := context.WithCancel(ctx)
+	c.ctx = ctx
+	c.cancel = cancel
+
+	go func() {
+		ticker := time.NewTicker(c.collectInterval)
+
+		for {
+			if err := c.extractSchema(c.ctx); err != nil {
+				level.Error(c.logger).Log("msg", "collector stopping due to error", "err", err)
+				c.Stop()
+				break
+			}
+
+			select {
+			case <-c.ctx.Done():
+				return
+			case <-ticker.C:
+				// continue loop
+			}
+		}
+	}()
+
+	return nil
+}
+
+func (c *SchemaTable) Stop() {
+	c.cancel()
+	c.dbConnection.Close()
```

Comment:
I don't think that the SchemaTable should be responsible to close the connection that's given as an arg. Since that the collectors are sharing the connections,  if one closes it because of an error it will be closed for all collectors?

Reply:
Good point, initially I was passing `dbConnection` to the collectors, but now that `dbConnection` is handled by the component I've moved `dbConnection.Close()` there.